fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockTest.java

```diff
@@ -130,6 +131,46 @@ void testFileLockCallback() throws IOException {
         }
     }
 
+    @Test
+    void testFileLockCallbackFrequently() throws IOException {
+        // This test simulates the situation where the AgilityContext flushes as part of the call to clear
+        final String lockName = "file.lock";
+        try (FDBRecordContext context = fdb.openContext()) {
+            AgilityContext agilityContext = AgilityContext.agile(context, 0, 0);
+            try (FDBDirectory directory = createDirectory(agilityContext)) {
+                final Lock lock1 = directory.obtainLock(lockName);
+                lock1.close();
+            }
+            agilityContext.flushAndClose();
+        }
+
+        assertCanObtainLock(lockName);
+    }
+
+    @ParameterizedTest
+    @BooleanSource
+    void testFileLockCallbackFrequentlyLost(boolean clearLock) throws IOException {
+        // This test simulates the situation where the AgilityContext flushes as part of the call to clear
+        final String lockName = "file.lock";
+        try (FDBRecordContext context = fdb.openContext()) {
+            AgilityContext agilityContext = AgilityContext.agile(context, 0, 0);
+            FDBDirectory directory = createDirectory(agilityContext);
+            try {
+                final Lock lock1 = directory.obtainLock(lockName);
+                if (clearLock) {
+                    forceClearLock(lockName);
+                } else {
+                    forceStealLock(lockName);
+                }
+                assertThrows(AlreadyClosedException.class, lock1::close);
+            } finally {
+                assertThrows(AlreadyClosedException.class, directory::close);
```

Comment:
I'm not entirely sure if this is the behavior we want here. 
Note: the `directory.close()` only fails if the lock is stolen above, if it is cleared, it won't throw an exception.

Reply:
Perhaps the lock should throw it, but the directory should not, at least from the javadoc for Lock:
```
   * Releases exclusive access.
   * <p>
   * Note that exceptions thrown from close may require
   * human intervention, as it may mean the lock was no
   * longer valid, or that fs permissions prevent removal
   * of the lock file, or other reasons.
   * <p>
 ```

Reply:
I agree that both stolen and cleared locks should cause an exception during an explicit lock close - as it means that Lucene assumes it has the lock while it doesn't.
The directory only calls `fileLockClearIfLocked` during close, so the scenario may be:

Getting an IO exception that leaves the lock behind
Closing the directory
Something else had stolen or cleared
An exception would not help in this recovery path, but maybe an error message will be appropriate if the lock was less than `timeWindowMilliseconds` ms old (again, for both stolen and cleared).